md4c/md4c.c

```diff
@@ -2331,6 +2331,23 @@ md_is_inline_link_spec(MD_CTX* ctx, const MD_LINE* lines, int n_lines,
     return ret;
 }
 
+static int
```

Comment:
I also have a trouble to understand why this function is needed at all. The caller `md_resolve_links()` already gets position of opening `[` and closing `]`. So the rescanning of all the contents between is a work in a vain. And could even cause troubles if you analyze `[[ [[ ]] ]]` because the scanner in this function will stop too early.

Reply:
I tried to follow the pattern of `md_is_inline_link_spec`. In it, there is a check to make sure that the link target does not span two lines, which is the exact case used for wiki link targets. So, I thought it would be suitable. And also, when support for the wiki link label is added, this function may include further checks, like preventing the label from spanning too many lines.

Reply:
The (ordinary) link handling is one of the most difficult parts of MD4C. The complexity is to avoid non-linear behavior for crazy malicious inputs. 

That's why there is that craziness with the `md_resolve_links()` at all. Without it it would be simple to DoS it with
```
$ python -c 'N=20000; print("[" * N + "foo" + "]" * N);' | ./md2html
```

I'm afraid your scanning in this function introduces the quadratic complexity with `--fwiki-links` again, and also that it is likely broken anyway because if you feed it with something like

```
[[ foo[[ bar]]
]]
```

Then for the outer `[[...]]` you have the new line there anyway but the scanner only always scans to the 1st `]]`.

Reply:
> The (ordinary) link handling is one of the most difficult parts of MD4C. The complexity is to avoid non-linear behavior for crazy malicious inputs.
> 
> That's why there is that craziness with the `md_resolve_links()` at all. Without it it would be simple to DoS it with
> 
> ```
> $ python -c 'N=20000; print("[" * N + "foo" + "]" * N);' | ./md2html
> ```

I see.

> I'm afraid your scanning in this function introduces the quadratic complexity with `--fwiki-links` again...

That's unfortunate. I will have another look at it and see if I can do something about it.

> ... and also that it is likely broken anyway because if you feed it with something like
> 
> ```
> [[foo[[bar]]
> ]]
> ```
> 
> Then for the outer `[[...]]` you have the new line there anyway but the scanner only always scans to the 1st `]]`.

This is the output that I get:

```
<p>[[foo<x-wikilink data-target="bar">bar</x-wikilink>
]]</p>
```

But that is what I expect:

- Links (and wiki links) are scanned from inside to out,
- Wiki link target cannot span more than one line

Reply:
> ... and also that it is likely broken anyway because if you feed it with something like
> ```
> [[foo[[bar]]
> ]]
> ```
> 
> Then for the outer `[[...]]` you have the new line there anyway but the scanner only always scans to the 1st `]]`.

Excuse me; maybe I was a bit too quick in my previous response. Do you mean that the following should be expected?
 
```
<p><x-wikilink data-target="foo[[bar">foo[[bar</x-wikilink>
]]</p>
```

Reply:
I think I have a proposal which would fix all of that:

Change the `md_is_inline_wikilink_spec` in the follow way:
1. Make it take `beg` and `end` offsets from the opening/closing `MD_MARK` so we really use the proper marks.
2. Make it scan backwards from `end`. It will come handy later when implementing the pipe case (we can then only scan over the ID, and keep the text without length limit).
3. Make it scan until it reaches `beg` instead of checking for `[[` because there may be nested `[[ ... ]]`.
4. Make it fail if it sees a forbidden char (`'\r'` or `'\n'` - here, it should be fine to use instead of analyzing the `lines[]`).
5. Make it scan at most, say 100 characters, and fail if it does not reach `beg`. (I think we can limit the ID length right?)

And I would also rename the function to `md_is_inline_wikilink_id()`. The `_spec()` for ordinary link is used because it is for the whole link destination + link title pair.

Reply:
> Excuse me; maybe I was a bit too quick in my previous response. Do you mean that the following should be expected?

Not necessarily. 

> But that is what I expect...

Maybe I constructed bad input exaple. 

The culprit of my message is that `md_resolve_links()` is called for each pair of `[` and `]` (following the nesting logic that inside you may get only balanced pairs), but your scanner will stop on the first `]]` so it has **a different logic** and does not necessarily scans whole `[[...]]` range as analyzed in the `md_resolve_links()`.

Reply:
> I think I have a proposal which would fix all of that:
> 
> Change the `md_is_inline_wikilink_spec` in the follow way:
> 
>     1. Make it take `beg` and `end` offsets from the opening/closing `MD_MARK` so we really use the proper marks.
> 
>     2. Make it scan backwards from `end`. It will come handy later when implementing the pipe case (we can then only scan over the ID, and keep the text without length limit).
> 
>     3. Make it scan until it reaches `beg` instead of checking for `[[` because there may be nested `[[ ... ]]`.
> 
>     4. Make it fail if it sees a forbidden char (`'\r'` or `'\n'` - here, it should be fine to use instead of analyzing the `lines[]`).
> 
>     5. Make it scan at most, say 100 characters, and fail if it does not reach `beg`. (I think we can limit the ID length right?)
> 
> 
> And I would also rename the function to `md_is_inline_wikilink_id()`. The `_spec()` for ordinary link is used because it is for the whole link destination + link title pair.

I did this; the result is much simpler. Thanks for the good advice.

Reply:
Seems to fix my counter-examples from above so seems to be on the right track now.